### PR TITLE
KEP-4818: Relaxed validation for allowing zero in PreStop hook sleep action

### PR DESCRIPTION
**Pull request by sreeram-venkitesh**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR adds a feature gate, `PodLifecycleSleepActionAllowZero`, to allow zero value for the sleep action in the PreStop container lifecycle hook.

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/enhancements/issues/4818

#### Does this PR introduce a user-facing change?

```release-note
Allows PreStop lifecycle handler's sleep action to have a zero value
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4818
```

### CODE REVIEW

File: pkg/apis/core/validation/validation.go
Comment (kannon92) [inline pkg/apis/core/validation/validation.go:3170]: We aim to not have feature gates in the validation logic.

You could use the feature gate in the PodValidationOpts.

**Comment by sreeram-venkitesh** [comment]
@kannon92 Addressing your comment, I've updated the logic to use the feature gate set in `PodValidationOptions`. I've tried to describe the change in my second commit with the flow given below. The validation takes place in `validateSleepAction`, for which I've passed the PodValidationOptions down from `validatePodSpec`, through the different functions which ultimately end up calling `validateHandler` and thus `validateSleepAction`.

```
validatePodSpec
|__validateContainers
|__validateInitContainers
   |__validateLifecycle
   |    |__validateHandler
   |    |__validateSleepAction
   |__validateLivenessProbe
   |	  |__validateProbe
   |	     |__validateHandler
   |	        |__validateSleepAction
   |__validateReadinessProbe
   |	  |__validateProbe
   |	     |__validateHandler
   |	        |__validateSleepAction
   |__validateStartupProbe
   	  |__validateProbe
   	     |__validateHandler
   	        |__validateSleepAction
```

**Comment by sreeram-venkitesh** [comment]
I'm working on fixing the tests.

**Comment by kannon92** [comment]
/hold

Please open a KEP first. You can leave this up but don't expect much reviews until the KEP is approved.

File: pkg/api/pod/util.go
Comment (thockin) [inline pkg/api/pod/util.go:389]: The usual pattern is to initialize this, then (below) do something like:

```
opts.AllowPodLifecycleSleepActionZeroValue = opts.AllowPodLifecycleSleepActionZeroValue || lifecycleSleepZeroUsed(podSpec)
```

Where:

```
func lifecycleSleepZeroUsed(spec *api.PodSpec) bool {
    for i := range spec.containers {
        c := &spec.Containers[i]
        if c.Lifecycle != nil {
            if c.Lifecycle.PostStart != nil && c.Lifecycle.PostStart.Sleep != nil && c.Lifecycle.PostStart.Sleep.Seconds == 0 {
                return true
            }
            if c.Lifecycle.PreStop != nil && c.Lifecycle.PreStop.Sleep != nil && c.Lifecycle.PreStop.Sleep.Seconds == 0 {
                return true
            }
        }
    }
    return false
}
```

There's a util function somewhere, I can't recall the name, that lets you easily repeat this for spec.Containers, .EphemeralContainers, and .InitContainers.  You should use that.

This allows us to respect existing pods which have set this value.  See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md for more

File: pkg/features/kube_features.go
Comment (thockin) [inline pkg/features/kube_features.go:606]: Also need it in pkg/features/versioned_kube_features.go

**Review by thockin — COMMENTED** [review]
This is nice, but it needs a unit test for validation and an e2e test for 0

File: pkg/api/pod/util.go
Comment (sreeram-venkitesh) [reply pkg/api/pod/util.go:389]: I believe the util function is [visitContainers](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L63).
https://github.com/kubernetes/kubernetes/blob/f153edf356324564396ae3f20538859783aadc76/pkg/api/pod/util.go#L63

It makes sense how by setting the following line `opts.AllowPodLifecycleSleepActionZeroValue = opts.AllowPodLifecycleSleepActionZeroValue || lifecycleSleepZeroUsed(podSpec)`, we can respect existing Pods with the value set to zero and make updates work and tolerate the zero value like you commented in the KEP Google doc. 

I can see a function with a similar pattern for the parent KEP too. This is really cool!
https://github.com/kubernetes/kubernetes/blob/f153edf356324564396ae3f20538859783aadc76/pkg/api/pod/util.go#L776-L781

I made the changes to add the `lifecycleSleepZeroUsed` function. I'm trying to test it locally but kind is throwing some errors. Let me take a look. I will push the code changes here as well as the KEP PR with the changes soon.

**Comment by sreeram-venkitesh** [comment]
@thockin PTAL! I've made it so that Pods with zero sleep duration can be updated and still continue to work and after the feature gate has been disabled. Also added some basic tests. 

CC: @SergeyKanzhelev @AxeZhan

File: test/e2e/common/node/lifecycle_hook.go
Comment (SergeyKanzhelev) [inline test/e2e/common/node/lifecycle_hook.go:633]: were there any tests to disable feature gate in scope of a KEP?

Comment (sreeram-venkitesh) [reply test/e2e/common/node/lifecycle_hook.go:633]: > a test with feature.PodLifecycleSleepActionAllowZero disabled, for example, when fg is disabled, creating a pod with 0 duration will result in a error?

Ah yes, I have added a test case in `validation_test.go` for this:
https://github.com/kubernetes/kubernetes/blob/a0d4262276ee5db42e9104f7640d6d800c9e73d3/pkg/apis/core/validation/validation_test.go#L24180-L24188

Comment (SergeyKanzhelev) [reply test/e2e/common/node/lifecycle_hook.go:633]: ok, the feature is simple enough to not have an e2e_node test for it.

File: pkg/api/pod/util.go
Comment (pacoxu) [inline pkg/api/pod/util.go:806]: Lifecycle is not allowed for ephemeral containers.
- However, `c.Lifecycle == nil` will return true for ephemeral containers.

File: pkg/apis/core/validation/validation.go
Comment (pacoxu) [inline pkg/apis/core/validation/validation.go:3184]: Should we include a suggestion to include that you can enable the new FG to allow zero in the error message?

File: pkg/apis/core/validation/validation.go
Comment (pacoxu) [inline pkg/apis/core/validation/validation.go:3178]: If gracePeriod can be nil, this should be
```
		if gracePeriod != nil && (sleep.Seconds < 0 || sleep.Seconds > *gracePeriod) {
```

File: pkg/apis/core/validation/validation_test.go
Comment (pacoxu) [inline pkg/apis/core/validation/validation_test.go:24219]: All gracePeriod here are not nil in the UT.

File: pkg/apis/core/validation/validation.go
Comment (pacoxu) [inline pkg/apis/core/validation/validation.go:3183]: same as above.

File: pkg/api/pod/util.go
Comment (sreeram-venkitesh) [reply pkg/api/pod/util.go:806]: Yes, for ephemeral containers `podLifecycleSleepActionZeroValueInUse` will return false after exiting from `VisitContainers`.

**Comment by sreeram-venkitesh** [comment]
@pacoxu I've added the changes from your review. For https://github.com/kubernetes/kubernetes/pull/127094#discussion_r1780552555 I believe it is fine for ephemeral containers since `podLifecycleSleepActionZeroValueInUse` returns false. This is based on the `podLifecycleSleepActionInUse` method which has the same behaviour. 

PTAL! Thank you!

**Comment by pacoxu** [comment]
> /hold
> 
> Please open a KEP first. You can leave this up but don't expect much reviews until the KEP is approved.

/unhold 
as the KEP is merged

/priority important-soon
/triage accepted

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [inline pkg/features/kube_features.go:580]: have we discussed skipping alpha on this KEP? Just a side note

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

e2e test is the best we can do I think

File: pkg/features/kube_features.go
Comment (sreeram-venkitesh) [reply pkg/features/kube_features.go:580]: Tim had mentioned doing one alpha release and then moving to beta/GA in [this comment](https://github.com/kubernetes/enhancements/issues/3960#issuecomment-2209320166) in the parent KEP.

File: pkg/features/kube_features.go
Comment (pacoxu) [inline pkg/features/kube_features.go:577]: We may remove this line as https://github.com/kubernetes/kubernetes/pull/128076.

**Comment by pacoxu** [comment]
> This is nice, but it needs a unit test for validation and an e2e test for 0

Ping @thockin to take another look for api review as your comments should be addressed.

**Comment by sreeram-venkitesh** [comment]
@pacoxu I've updated the kube_features.go file as per https://github.com/kubernetes/kubernetes/pull/128076 and the lgtm has been removed hence.

File: test/e2e/feature/feature.go
Comment (SergeyKanzhelev) [inline test/e2e/feature/feature.go:271]:
```suggestion
	// Marks a single test that tests Pod Lifecycle Sleep action with zero duration. Requires feature gate PodLifecycleSleepActionAllowZero to be enabled.
```

**Review by SergeyKanzhelev — APPROVED** [review]
nothing much from the node side. And already approved. API side changes look reasonable.

Please squash commits before lgtm

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Review by hashim21223445 — APPROVED** [review]

**Review by hashim21223445 — APPROVED** [review]

File: pkg/features/kube_features.go
Comment (thockin) [reply pkg/features/kube_features.go:463]: Need to be able to roll-back, which would make saved data invalid.  Alpha->GA for this seems fine.

**Comment by sreeram-venkitesh** [comment]
Thanks Tim!

@SergeyKanzhelev lgtm was removed after rebase

**Pull request merged**
